wpilibc/src/main/native/include/frc/smartdashboard/SendableChooser.inc

```diff
@@ -98,6 +104,10 @@ void SendableChooser<T>::InitSendable(nt::NTSendableBuilder& builder) {
                               std::scoped_lock lock(m_mutex);
                               m_haveSelected = true;
                               m_selected = val;
+                              if (m_previousVal != val) {
+                                m_listener(GetSelected());
```

Comment:
GetSelected() also tries to lock m_mutex.  wpi::mutex is non-recursive, so this won't work.  Since you already have the value and you just made it selected, why can't this simply be `m_listener(val)`?

Reply:
Should the listener consume the string key or the value attached to the key? That's why I'm calling `GetSelected`, to get the value attached to the key.

Reply:
Good point.  It should be `m_listener(m_choices[val])` then.

Reply:
Although to not hold the lock, you'll have to do something like `auto choice = m_choices[val]; auto listener = m_listener; ... release lock ... ; listener(choice);`

Reply:
Do I need to re-lock it again? If so, how would I do that?

Reply:
No need to re-lock if you're just exiting the function.  To simplify things, I would put the callback code at the end of the function--so you exit early if m_listener is null, and then copy and unlock, then exit.

Reply:
How do I unlock the mutex? `scoped_lock` doesn't seem to have an unlock, and `m_mutex.unlock()` also didn't work, errors out with
```
D:\a\_work\1\s\src\vctools\crt\github\stl\src\mutex.cpp(164): unlock of unowned mutex
D:\a\_work\1\s\src\vctools\crt\github\stl\src\mutex.cpp(61): mutex destroyed while busy
```

Reply:
You can limit the scope with curly braces.